Builder methods for remaining GLSLstd450 instructions

trenouf opened the pull request:
This adds builder methods for the remaining GLSLstd450 instructions, and
switches the SPIR-V translator to use them.

Change-Id: I2b58ae2bd7fcdb43869b4c200ef60431a0c22836

trenouf commented:
Please review only topmost commit.

linqun commented on translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -9572,9 +9747,83 @@ Value *SPIRVToLLVM::transGLSLExtInst(SPIRVExtInst *ExtInst,
+  case GLSLstd450FindSMsb: {
+    // Find signed integer most-significant bit.
+    // The spec claims that the result must be the same type as the input, but I
+    // have seen SPIR-V that does not do that.
```
- Where did you find the SPIR-V binary? games or CTS? Is it possible to add it as comment?
- What's the validation result with spirv-tools?

trenouf commented:
Testing has shown up a few bugs in this, and I am changing parameter names in line with earlier reviews. So it is not worth reviewing this one until I have pushed a new version of it.

linqun commented on translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -9510,11 +9687,10 @@ Value *SPIRVToLLVM::transGLSLExtInst(SPIRVExtInst *ExtInst,
     Value *Multiplier =
         getBuilder()->CreateFDiv(ConstantFP::get(Val->getType(), 1.0),
-                               ConstantFP::get(Val->getType(), 32767.0));
+                                 ConstantFP::get(Val->getType(), 32767.0));
```
please remove unnecessary CreateFDiv in this comment

trenouf replied:
I have resolved this in #168.

linqun commented on translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -9441,15 +9478,155 @@ Value *SPIRVToLLVM::transGLSLExtInst(SPIRVExtInst *ExtInst,
+  }
+
+  case GLSLstd450FMin:
+  case GLSLstd450NMin: {
```
we need support nan for NMin

trenouf replied:
Good spot. I had already found that in a CTS failure and fixed it locally, but I forgot to fix nmax in the same way. :-)

linqun commented on translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -9329,31 +9329,67 @@ Instruction *SPIRVToLLVM::transOCLBuiltinFromExtInst(SPIRVExtInst *BC,
+
+  case GLSLstd450FSign:
+    // Get sign of FP value
+    return getBuilder()->CreateFSign(Args[0]);
```
please check whether it breaks CTS float_control tests

trenouf replied:
This does not break those tests, but I found several other problems from those tests that I have now fixed.

linqun commented on translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -9441,15 +9478,155 @@ Value *SPIRVToLLVM::transGLSLExtInst(SPIRVExtInst *ExtInst,
+
+  case GLSLstd450Fma:
+    // Fused multiply and add
+    return getBuilder()->CreateIntrinsic(Intrinsic::fmuladd, Args[0]->getType(),
```
please use CreateFma

linqun commented on translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -9441,15 +9478,155 @@ Value *SPIRVToLLVM::transGLSLExtInst(SPIRVExtInst *ExtInst,
+  case GLSLstd450NMax: {
+    // FP maximum
+    FastMathFlags FMF;
+    FMF.setNoNaNs();
```
we need nan for NMax

linqun commented on translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -9441,15 +9478,155 @@ Value *SPIRVToLLVM::transGLSLExtInst(SPIRVExtInst *ExtInst,
+  case GLSLstd450FClamp:
+  case GLSLstd450NClamp:
+    // FP clamp
+    return getBuilder()->CreateFMed3(Args[0], Args[1], Args[2]);
```
please set nan flag for GLSLstd450FClamp

trenouf replied:
Actually I created a new CreateFClamp to use here, and set FMF appropriately for it depending on whether it is FClamp or NClamp. I discovered that using CreateFMed3 is bad for a datatype that does not support the fmed3 instruction, because it does 2x min and 2x max, whereas for a clamp you only need 1x min and 1x max.

trenouf commented:
V2: Rebase
    V3: Fixed normalize scalar.
    V4: Rebased.
    V5: Fixed nmin NaN behavior. Fixed reflect and refract param order.
        Anticipated param name review comments. Added fclamp to builder,
        including setting FMF in BuilderReplayer
    V6: Rebased, removing accidental change to pack/unpack code.
    V7: Fixes:
        - Fixed nan handling on nmax and nmin.
        - Fixed sstep.
        - Don't use fmed3 for nclamp.
        - Use fma inst for spir-v fma on gfx9.
        - More param name fixes.
        - Rebased.

This is still two commits, awaiting a rebase on the merge of #168. Please review just the topmost commit.

trenouf commented:
V8: More fixes:
        - Fixed FindUMsb and FindLsb.
        - Fixed param order in reflect.

amdrexu commented:
When this commit is rebased, please let me know. I will check it. There are more codes in this commit, please let me have one day to look through it. Thanks.

amdrexu commented on builder/llpcBuilderImplArith.cpp:
```diff
@@ -510,6 +528,17 @@ Value* BuilderImplArith::CreateInverseSqrt(
+    Value*        pX,         // [in] Input value
+    const Twine&  instName)   // [in] Name to give instruction(s)
+{
+    Value* pNegated = CreateSub(Constant::getNullValue(pX->getType()), pX);
```
I think you can use this CreateNeg()

pNegated -> pNegX

amdrexu commented on builder/llpcBuilderImplArith.cpp:
```diff
@@ -523,6 +552,251 @@ Value* BuilderImplArith::CreateFSign(
+{
+    Value* pDiff = CreateFSub(pX, pEdge0);
+    Constant* pOne = ConstantFP::get(pX->getType(), 1.0);
+    Value* pTValue = CreateFMul(pDiff, CreateFDiv(pOne, CreateFSub(pEdge1, pEdge0)));
```
pTValue -> pT

amdrexu commented on builder/llpcBuilderImplArith.cpp:
```diff
@@ -756,3 +1031,26 @@ Value* BuilderImplArith::CreateExtractBitField(
+// Create "find MSB" operation for a (vector of) signed i32. For a postive number, the result is the bit number of
+// the most significant 1-bit. For a negative number, the result is the bit number of the most significant 0-bit.
+// For a value of 0 or -1, the result is -1.
+Value* BuilderImplArith::CreateFindSignedMSB(
```
CreateFindSignedMSB -> CreateFindSMsb

amdrexu commented on builder/llpcBuilderImplArith.cpp:
```diff
@@ -756,3 +1031,26 @@ Value* BuilderImplArith::CreateExtractBitField(
+// the most significant 1-bit. For a negative number, the result is the bit number of the most significant 0-bit.
+// For a value of 0 or -1, the result is -1.
+Value* BuilderImplArith::CreateFindSignedMSB(
+    Value*        pInValue,   // [in] Input value
```
pInValue -> pValue

amdrexu commented on builder/llpcBuilderImplArith.cpp:
```diff
@@ -756,3 +1031,26 @@ Value* BuilderImplArith::CreateExtractBitField(
+{
+    LLPC_ASSERT(pInValue->getType()->getScalarType()->isIntegerTy(32));
+
+    Constant* pMinus1 = ConstantInt::get(pInValue->getType(), -1);
```
pMinus1 -> pNegOne

amdrexu commented on builder/llpcBuilderImplArith.cpp:
```diff
@@ -756,3 +1031,26 @@ Value* BuilderImplArith::CreateExtractBitField(
+                              return CreateIntrinsic(Intrinsic::amdgcn_sffbh, pInValue->getType(), pInValue);
+                           });
+    Value* pBitp = CreateSub(ConstantInt::get(pInValue->getType(), 31), pLz);
+    Value* pIsNeg1 = CreateICmpEQ(pInValue, pMinus1);
```
pIsNeg1 -> pIsNegOne

amdrexu commented on builder/llpcBuilderImplArith.cpp:
```diff
@@ -756,3 +1031,26 @@ Value* BuilderImplArith::CreateExtractBitField(
+    LLPC_ASSERT(pInValue->getType()->getScalarType()->isIntegerTy(32));
+
+    Constant* pMinus1 = ConstantInt::get(pInValue->getType(), -1);
+    Value* pLz = Scalarize(pInValue,
```
pLz -> pLeadingZeroCount

amdrexu commented on builder/llpcBuilderImplArith.cpp:
```diff
@@ -756,3 +1031,26 @@ Value* BuilderImplArith::CreateExtractBitField(
+                           {
+                              return CreateIntrinsic(Intrinsic::amdgcn_sffbh, pInValue->getType(), pInValue);
+                           });
+    Value* pBitp = CreateSub(ConstantInt::get(pInValue->getType(), 31), pLz);
```
pBitp -> pBitOnePos

amdrexu commented on builder/llpcBuilderImplArith.cpp:
```diff
@@ -756,3 +1031,26 @@ Value* BuilderImplArith::CreateExtractBitField(
+    Value* pBitp = CreateSub(ConstantInt::get(pInValue->getType(), 31), pLz);
+    Value* pIsNeg1 = CreateICmpEQ(pInValue, pMinus1);
+    Value* pIsZero = CreateICmpEQ(pInValue, Constant::getNullValue(pInValue->getType()));
+    Value* pResIsNeg1 = CreateOr(pIsNeg1, pIsZero);
```
pResIsNeg1 -> pIsNegOneOrZero

trenouf commented:
V9: Rebased. Added and used CreateFMin and CreateFMax, which also
    encapsulate pre-gfx9 denorm folding when that mode is set.
V10: Rebased on FMF changes.
V11: Addressed review comments.

trenouf commented:
V12: Disabled FMF on f16 smoothstep.

amdvlk-admin commented:
### Test summary for commit 3cdfa0edcb2082a277415284281e1b9fd7f3dc9f
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev dcb42575c3d374467bb6e4e609f8bf6d1302c297</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>
